compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -839,38 +941,72 @@ impl<'a> CoverageSpanRefinery<'a> {
-        // other branches).
-
-        self.to_refined_spans_without_closures()
-    }
+        refined_spans.push(prev);
+
+        // Remove `CoverageSpan`s with empty spans ONLY if the empty `CoverageSpan`s BCB also has at
+        // least one other non-empty `CoverageSpan`.
+        let mut has_coverage = BitSet::new_empty(basic_coverage_blocks.num_nodes());
+        for covspan in &refined_spans {
+            if !covspan.span.is_empty() {
+                has_coverage.insert(covspan.bcb_leader_bb);
+            }
+        }
+        refined_spans.retain(|covspan| {
+            !(covspan.span.is_empty()
+                && is_goto(&mir_body[covspan.bcb_leader_bb].terminator().kind)
+                && has_coverage.contains(covspan.bcb_leader_bb))
+        });
 
-    fn add_refined_span(&mut self, coverage_span: CoverageSpan) {
-        self.refined_spans.push(coverage_span);
+        // Remove `CoverageSpan`s derived from closures, originally added to ensure the coverage
+        // regions for the current function leave room for the closure's own coverage regions
+        // (injected separately, from the closure's own MIR).
+        refined_spans.retain(|covspan| !covspan.is_closure);
+        refined_spans
     }
 
-    /// Remove `CoverageSpan`s derived from closures, originally added to ensure the coverage
-    /// regions for the current function leave room for the closure's own coverage regions
-    /// (injected separately, from the closure's own MIR).
-    fn to_refined_spans_without_closures(mut self) -> Vec<CoverageSpan> {
-        self.refined_spans.retain(|covspan| !covspan.is_closure);
-        self.refined_spans
+    // Generate a set of `CoverageSpan`s from the filtered set of `Statement`s and `Terminator`s of
+    // the `BasicBlock`(s) in the given `BasicCoverageBlock`. One `CoverageSpan` is generated
+    // for each `Statement` and `Terminator`. (Note that subsequent stages of coverage analysis will
+    // merge some `CoverageSpan`s, at which point a `CoverageSpan` may represent multiple
+    // `Statement`s and/or `Terminator`s.)
+    fn bcb_to_initial_coverage_spans(&self, bcb: &BasicCoverageBlock) -> Vec<CoverageSpan> {
+        bcb.blocks
+            .iter()
+            .map(|bbref| {
+                let bb = *bbref;
+                let data = &self.mir_body[bb];
+                data.statements
+                    .iter()
+                    .enumerate()
+                    .filter_map(move |(index, statement)| {
+                        filtered_statement_span(statement, self.body_span).map(|span| {
+                            CoverageSpan::for_statement(statement, span, bcb, bb, index)
+                        })
+                    })
+                    .chain(
+                        filtered_terminator_span(data.terminator(), self.body_span)
+                            .map(|span| CoverageSpan::for_terminator(span, bcb, bb)),
+                    )
+            })
+            .flatten()
```

Comment:
nit: I think you can use `flat_map` instead of `map` then `flatten`?

Reply:
👍 Done